### PR TITLE
pg_tap

### PR DESCRIPTION
**Pull request by outscale-jju**

### CODE REVIEW

File: src/tap.c [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+		pg_brick_get_state(brick, struct pg_tap_state);
+
+	return state->ifr.ifr_name;
+}
+
+static struct pg_brick_config *tap_config_new(const char *name,
+					      const char *ifname)
+{
+	struct pg_brick_config *config = g_new0(struct pg_brick_config, 1);
+	struct pg_tap_config *tap_config = g_new0(struct pg_tap_config, 1);
```
Comment (outscale-mgo) [inline src/tap.c]: config->brick_config = g_new0(struct pg_tap_config, 1);

File: src/tap.c [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	struct pg_tap_state *state = pg_brick_get_state(brick,
+							struct pg_tap_state);
+	uint64_t it_mask;
+	uint64_t bit;
+	uint16_t i;
+	int fd = state->tap_fd;
+
+	it_mask = pkts_mask;
+	for (; it_mask;) {
+		pg_low_bit_iterate_full(it_mask, bit, i);
```
Comment (outscale-mgo) [inline src/tap.c]: if you don;t use bit, you can use pg_low_bit_iterate instead of pg_low_bit_iterate_full

File: src/tap.c [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+	uint16_t nb_pkts;
+	uint64_t pkts_mask;
+	int ret;
+	struct pg_tap_state *state =
+		pg_brick_get_state(brick, struct pg_tap_state);
+	int fd = state->tap_fd;
+	int fd1 = fd + 1;
+	struct pg_brick_side *s = &brick->side;
+	struct rte_mbuf **pkts = state->pkts;
+	static struct timeval timeout;
```
Comment (outscale-mgo) [inline src/tap.c]: I would add a comment to explain that by having a static timeout struct, we're implicitly set the struct to 0 therefore we're aren't waiting in select.

**Comment by outscale-mgo** [comment]
core: move mempool defines in header should be merge with the 1rst commit.

File: src/tap.c [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+		}
+
+		if (!FD_ISSET(fd, &fdset))
+			break;
+
+		struct rte_mbuf *packet = pkts[nb_pkts];
+
+		rte_pktmbuf_reset(packet);
+		char *data = rte_pktmbuf_mtod(packet, char *);
+		int read_size = read(fd, data, MBUF_SIZE);
```
Comment (outscale-mgo) [inline src/tap.c]: what if you have a packet bigger than a MBU_SIZE ?

File: src/tap.c [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+	}
+
+	state = pg_brick_get_state(brick, struct pg_tap_state);
+	memset(&state->ifr, 0, sizeof(struct ifreq));
+	tap_config = config->brick_config;
+	strncpy(state->ifr.ifr_name, tap_config->ifname, IFNAMSIZ);
+	state->ifr.ifr_flags = IFF_NO_PI | IFF_TAP;
+
+	if (ioctl(tap_fd, TUNSETIFF, (void *) &state->ifr) < 0) {
+		close(tap_fd);
```
Comment (outscale-mgo) [inline src/tap.c]: you could use a goto error:

File: src/tap.c [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+static void tap_destroy(struct pg_brick *brick, struct pg_error **errp)
+{
+	struct pg_tap_state *state =
+		pg_brick_get_state(brick, struct pg_tap_state);
+	struct rte_mbuf **pkts = state->pkts;
+
+	close(state->tap_fd);
+	for (int i = 0; i < PG_MAX_PKTS_BURST; i++)
```
Comment (outscale-mgo) [inline src/tap.c]: pg_packets_free

File: tests/tap/tests.c [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+	struct pg_error *error = NULL;
+	const char *ifname;
+
+	/* create tap */
+	g_assert(!iface_exists("YAY"));
+	tap = pg_tap_new("tap", "YAY", &error);
+	g_assert(tap);
+	g_assert(!error);
+	ifname = pg_tap_ifname(tap);
+	g_assert(ifname);
```
Comment (outscale-mgo) [inline tests/tap/tests.c]: we can remove this line because there's the exact same check in iface_exists

File: tests/tap/tests.c [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+	/* destroy tap */
+	pg_brick_destroy(tap);
+	g_assert(!iface_exists("YAY"));
+
+	/* re-create tap */
+	tap = pg_tap_new("tap", "YAY", &error);
+	g_assert(tap);
+	g_assert(!error);
+	ifname = pg_tap_ifname(tap);
+	g_assert(ifname);
```
Comment (outscale-mgo) [inline tests/tap/tests.c]: same here

File: tests/tap/tests.c [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+	g_assert(g_strcmp0("YAY", ifname) == 0);
+	g_assert(iface_exists("YAY"));
+
+	/* try to create an existing tap */
+	tap2= pg_tap_new("tap2", "YAY", &error);
+	g_assert(!tap2);
+	g_assert(error);
+	pg_error_free(error);
+	error = NULL;
+
```
Comment (outscale-mgo) [inline tests/tap/tests.c]: you should check that you can create 2 tap with different names

File: src/tap.c [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+		static_assert(PG_MBUF_SIZE > 1500,
+			      "reminder: should fix code below");
+		/* ignore potential truncated packets */
+		if (unlikely(read_size == PG_MBUF_SIZE)) {
+			nb_pkts--;
+			continue;
+		}
+
+		if (read_size < 0) {
```
Comment (outscale-mgo) [inline src/tap.c]: unlikely

File: src/tap.c [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+
+	it_mask = pkts_mask;
+	for (; it_mask;) {
+		pg_low_bit_iterate(it_mask, i);
+		struct rte_mbuf *packet = pkts[i];
+		char *data = rte_pktmbuf_mtod(packet, char *);
+		int len = rte_pktmbuf_pkt_len(packet);
+
+		/* write data. */
+		if (write(fd, data, len) < 0) {
```
Comment (outscale-mgo) [inline src/tap.c]: unlikely

**Comment by outscale-jju** [comment]
merged

**Pull request closed without merging**